pkg/volume/glusterfs/glusterfs.go

```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	}
+	// Collect the SC names and allocated GIDs from existing PVs.
+	scNameGids := make(map[string][]string)
+	for _, pv := range pvList.Items {
+		if pv.Annotations[annDynamicallyProvisioned] == glusterfsPluginName {
+			scNameInAnnot := pv.Annotations[annStorageClass]
+			scNameGids[scNameInAnnot] = append(scNameGids[scNameInAnnot], pv.Annotations[volumehelper.VolumeGidAnnotationKey])
+		}
+	}
+	glog.V(3).Infof("glusterfs: PVs with SC name and allocated GID Map :%v ", scNameGids)
+
+	for key := range scNameGids {
+		scdetails, _ := r.plugin.host.GetKubeClient().Storage().StorageClasses().Get(key)
+		if scdetails.Parameters["gidMin"] == "" {
+			scdetails.Parameters["gidMin"] = defaultGidMin
+		}
+		if scdetails.Parameters["gidMax"] == "" {
+			scdetails.Parameters["gidMax"] = defaultGidMax
+		}
+		blockStartnew32, err := r.ConvertGid(scdetails.Parameters["gidMin"])
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %s", scdetails.Parameters["gidMin"])
+		}
+		blockEndnew32, err := r.ConvertGid(scdetails.Parameters["gidMax"])
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %s", scdetails.Parameters["gidMax"])
+		}
+		// For each SC, create a pool and store it in plugin allocator map.
+		if r.plugin.gidAllocators[key] == nil {
+			r.plugin.gidAllocators[key], err = gidPoolCreate(blockStartnew32, blockEndnew32)
+			if err != nil {
+				glog.Errorf("glusterfs: gid pool creation failed %v", err)
+				return fmt.Errorf("glusterfs: gid pool creation failed: %v.", err)
+			}
+		}
+	}
+
+	// For each SCs in the allocator, set bitmap for existing/previosly allocated GIDs.
+	for name, gids := range scNameGids {
+		for _, gid := range gids {
+			tempGid, err := r.ConvertGid(gid)
+			if err == nil {
+				if r.plugin.gidAllocators[name] != nil {
+					err = r.plugin.gidAllocators[name].Allocate(Block{Start: tempGid, End: tempGid})
+				}
+			} else {
+				glog.Errorf("glusterfs: failed to convert gid %s", gid)
+			}
+		}
+	}
+	return nil
+}
+
 func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
 	var err error
+	// Controller restart is handled here with initialized bool in plugin struct.
+	if !r.plugin.initialized {
+		err = r.ConstructBitmap()
+		if err != nil {
+			glog.Errorf("glusterfs: failed to construct allocators from existing PVs")
+
+		}
+		r.plugin.initialized = true
+	}
+
+	glog.V(1).Infof("glusterfs: allocators after restoring: %#+v ", r.plugin.gidAllocators)
+
+	var blockStart32, blockEnd32 uint32
 	if r.options.PVC.Spec.Selector != nil {
 		glog.V(4).Infof("glusterfs: not able to parse your claim Selector")
 		return nil, fmt.Errorf("glusterfs: not able to parse your claim Selector")
 	}
 	glog.V(4).Infof("glusterfs: Provison VolumeOptions %v", r.options)
-
 	cfg, err := parseClassParameters(r.options.Parameters, r.plugin.host.GetKubeClient())
 	if err != nil {
 		return nil, err
 	}
 	r.provisioningConfig = *cfg
```

Comment:
answering my own question, a new glusterfsVolumeProvisioner instance is constructed per PVC provision operation